core/src/main/java/io/grpc/AbstractServerBuilder.java

```diff
@@ -114,18 +111,39 @@ public ServerImpl build() {
       releaseExecutor = true;
     }
 
-    ServerImpl server = new ServerImpl(executor, registry);
-    server.setTransportServer(buildTransportServer(server.serverListener()));
+    final ServerEssentials essentials = buildEssentials();
+    ServerImpl server = new ServerImpl(executor, registry, essentials.server);
     server.setTerminationRunnable(new Runnable() {
       @Override
       public void run() {
         if (releaseExecutor) {
           SharedResourceHolder.release(DEFAULT_EXECUTOR, executor);
         }
+        if (essentials.terminationRunnable != null) {
+          essentials.terminationRunnable.run();
+        }
       }
     });
     return server;
   }
 
-  protected abstract Service buildTransportServer(ServerListener serverListener);
+  protected abstract ServerEssentials buildEssentials();
+
+  protected static class ServerEssentials {
+    final io.grpc.transport.Server server;
+    @Nullable
+    final Runnable terminationRunnable;
+
+    /**
+     * Constructor.
+     *
+     * @param server the created server uses this server to accept transports
+     * @param terminationRunnable will be called at the server termination
+     */
+    public ServerEssentials(io.grpc.transport.Server server,
```

Comment:
ServerEssentials itself is protected, so this is implicitly so. I purposefully don't propagate such visibility scoping down into all the usages, as then if we want to change the visibility we have to re-discover what each visibility should be and it would violate a DRY principal. In addition, if I see something "protected" here I have to look at the visibility of the class to decide if it matters or not.